core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -0,0 +1,780 @@
+
+    } else if (stateOfRemovedAug == State.VISIBLE) {
+      // All segments in the visible atomicUpdateGroup which overshadows this atomicUpdateGroup is removed.
+      // Fall back if there is a fully available overshadowed atomicUpdateGroup
+
+      final List<AtomicUpdateGroup<T>> latestFullAugs = findLatestFullyAvailableOvershadowedAtomicUpdateGroup(
+          rangeOfAug,
+          minorVersion
+      );
+
+      if (!latestFullAugs.isEmpty()) {
+        // Move the atomicUpdateGroup to standby
+        // and move the fully available overshadowed atomicUpdateGroup to visible
+        if (!augOfRemovedChunk.isEmpty()) {
+          transitPartitionChunkState(augOfRemovedChunk, State.VISIBLE, State.STANDBY);
+        }
+        latestFullAugs.forEach(group -> transitPartitionChunkState(group, State.OVERSHADOWED, State.VISIBLE));
+      }
+    } else {
+      // do nothing
+    }
+  }
+
+  private List<AtomicUpdateGroup<T>> findLatestFullyAvailableOvershadowedAtomicUpdateGroup(
+      RootPartitionRange rangeOfAug,
+      short minorVersion
+  )
+  {
+    final List<Short2ObjectMap.Entry<AtomicUpdateGroup<T>>> overshadowedGroups = findOvershadowedBy(
+        rangeOfAug,
+        minorVersion,
+        State.OVERSHADOWED
+    );
+    if (overshadowedGroups.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    final OvershadowableManager<T> manager = new OvershadowableManager<>();
+    overshadowedGroups.stream()
+                      .flatMap(entry -> entry.getValue().getChunks().stream())
+                      .forEach(manager::add);
+
+    return manager.visibleGroup
+        .values()
+        .stream()
+        .flatMap(versionToGroup -> versionToGroup.values().stream())
+        .collect(Collectors.toList());
+  }
+
+  private void removeFrom(AtomicUpdateGroup<T> aug, State state)
+  {
+    final RootPartitionRange rangeKey = RootPartitionRange.of(aug);
+    final Short2ObjectSortedMap<AtomicUpdateGroup<T>> versionToGroup = getStateMap(state).get(rangeKey);
+    if (versionToGroup == null) {
+      throw new ISE("Unknown atomicUpdateGroup[%s] in state[%s]", aug, state);
+    }
+
+    final AtomicUpdateGroup<T> removed = versionToGroup.remove(aug.getMinorVersion());
+    if (removed == null) {
+      throw new ISE("Unknown atomicUpdateGroup[%s] in state[%s]", aug, state);
+    }
+
+    if (!removed.equals(aug)) {
+      throw new ISE(
+          "WTH? actually removed atomicUpdateGroup[%s] is different from the one which is supposed to be[%s]",
+          removed,
+          aug
+      );
+    }
+
+    if (versionToGroup.isEmpty()) {
+      getStateMap(state).remove(rangeKey);
+    }
+  }
+
+  @Nullable
+  public PartitionChunk<T> remove(PartitionChunk<T> partitionChunk)
+  {
+    final PartitionChunk<T> knownChunk = knownPartitionChunks.get(partitionChunk.getChunkNumber());
+    if (knownChunk == null) {
+      return null;
+    }
+
+    if (!knownChunk.equals(partitionChunk)) {
```

Comment:
If there are replicas, does this need to account for the number of replicas?

Reply:
`VisibleIntervalTimeline` doesn't handle replicas properly now. It should be handled before building timeline. See `timelines` in `BrokerServerView` as an example.